dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/infer/InferrerHelper.java

```diff
@@ -0,0 +1,318 @@
+            LOG.error(message);
+            throw new IllegalStateException(message);
+        }
+    }
+
+    /**
+     * Tweaks the specified entries so that that they have unique names and no duplicate names or paths.
+     */
+    public List<Inferrer.Entry> refine(List<Inferrer.Entry> entries) {
+        return ensureUniqueNames(legalizeNames(setMissingNames(removeDuplicatePaths(entries))));
+    }
+
+    private List<Inferrer.Entry> removeDuplicatePaths(List<Inferrer.Entry> entries) {
+        Set<String> paths = new HashSet<>();
+        return entries.stream().filter(entry -> paths.add(entry.path())).toList();
+    }
+
+    private List<Inferrer.Entry> setMissingNames(List<Inferrer.Entry> entries) {
+        return entries.stream().map(this::setMissingName).toList();
+    }
+
+    private Inferrer.Entry setMissingName(Inferrer.Entry entry) {
+        return entry.changeName(StringUtils.firstNonEmpty(entry.name(), nameFromPath(entry.path()), nameFromType(entry.type())));
+    }
+
+    private List<Inferrer.Entry> legalizeNames(List<Inferrer.Entry> entries) {
+        return entries.stream().map(this::legalizeName).toList();
+    }
+
+    private Inferrer.Entry legalizeName(Inferrer.Entry entry) {
+        String legalName = entry.name()
+            // Remove any character that's not an alphanumeric, hyphen, or underscore.
+            .replaceAll("[^a-zA-Z0-9_-]", "")
+            // Remove leading and trailing hyphens/underscores.
+            .replaceAll("^[_-]++", "")
+            .replaceAll("[_-]++$", "")
+            // Reduce remaining runs of hyphens/underscores to a single character.
+            .replaceAll("([_-])[_-]++", "$1");
+        return entry.changeName(StringUtils.firstNonEmpty(legalName, nameFromType(entry.type())));
+    }
+
+    private List<Inferrer.Entry> ensureUniqueNames(List<Inferrer.Entry> entries) {
+        // Compute the set of duplicate names.
+        Set<String> names = new HashSet<>();
+        Set<String> duplicates = new HashSet<>();
+        entries.forEach(entry -> {
+            String name = entry.name();
+            if (names.contains(name)) {
+                duplicates.add(name);
+            }
+            names.add(name);
+        });
+
+        // For each name, compute a new name if the name is either:
+        //   a) in the set of duplicate names.
+        //   b) has already been taken.
+        // Create new names by appending an underscore and the value of a counter to the
+        // original name.  For example, given duplicate name "a", we would try the names
+        // "a_1", "a_2", "a_3", ...
+        // Currently, a single counter is used across all names.
+        // In a perfect world, there would be a counter per orginal name, but this code is
+        // already hard to evaluate from a "does it always halt" point-of-view, so let's
+        // keep it simple for now.
+        AtomicLong counter = new AtomicLong(1);
```

Comment:
> Why is this an `AtomicLong`? I'm used to seeing Atomics being used in concurrency contexts, and I'm wondering if I'm missing something. Nothing wrong with it I suppose, but if there are concurrency concerns (which I don't see), then you would probably need to handle the various `Set`s as well.

Here, I use AtomicLong as a counter (mutable long), which will make it easier to adapt and keep the code simple if we someday use a Map of AtomicLong rather than a single one.  Yes, it could be a `long`, for now.

Reply:
Think we have https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/mutable/MutableLong.html on the classpath.